docs/langref/relay_type.rst

```diff
@@ -0,0 +1,213 @@
+===================
+Relay's Type System
+===================
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have not yet described the type system. Relay is
+a statically typed and type-inferred language, allowing programs to
+be typed with a minimal requirement of explicit type information.
+
+Static types are useful when performing compiler optimizations because they
+communicate properties about the data we manipulate, such as runtime shape,
+data layout, and storage without needing to run the program. Additionally,
+static typing is useful for determining the 
+
+Relay's type system features a form of depending typing for shapes to
+replace the dynamic shape inference used in NNVM and most other machine
+learning IRs. Treating tensor shapes as types allows Relay to perform
+more powerful reasoning at compile time; in particular, Relay can statically
+reason about operations whose output shapes vary based on the input shapes
+in complex ways. Statically reasoning about shapes in this manner allows
```

Comment:
Can you give a concrete example delineating the mechanism of this feature? It's important for people coming from the dataflow background, the NNVM way, to understand what is happening under the hood.

Reply:
This is handled through Relay's "pass" mechanism. I will add a sentence or two about it, but I think a full discussion of it might merit a document of its own. @jroesch, do you think we should have a page on writing Relay passes and how they work?